### PR TITLE
refactor tree (重构&优化资产树/用户授权树加载速度) (#5548)

### PR DESCRIPTION
**Pull request by BaiJiangJie**

### CODE REVIEW

File: apps/assets/models/node.py [hunk 1]
```diff
@@ -287,6 +295,20 @@ def get_nodes_all_assets_ids(cls, nodes_keys):
...
+    def get_nodes_all_assets_v2(cls, *nodes):
+        from .asset import Asset
+        node_ids = set()
+        descendant_node_query = Q()
+        for n in nodes:
+            node_ids.add(n.id)
+            descendant_node_query |= Q(key__istartswith=f'{n.key}:')
+        if descendant_node_query:
+            node_ids.update(
+                Node.objects.order_by().filter(descendant_node_query).values_list('id', flat=True)
```
Comment (ibuler) [inline apps/assets/models/node.py:308]: 拆出来

File: apps/assets/models/node.py [hunk 2]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
         if q:
-            return Asset.org_objects.filter(q).distinct()
+            return Asset.objects.filter(q).distinct()
         else:
             return Asset.objects.none()
 
+    # Use a new plan
+
+    # { org_id: { node_key: [ asset1_id, asset2_id ] } }
+    org_mapping_to_node_all_assets_id_mapping = defaultdict(dict)
```
Comment (ibuler) [inline apps/assets/models/node.py:335]: orgid_nodekey_assetsid_mapping = {}
orgId_nodeKey_assetsId_mapping

File: apps/assets/models/node.py [hunk 3]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
+        _assets_id = self.get_all_assets_id()
+        return len(_assets_id)
+
+    def get_all_assets_id(self):
+        _assets_id = self.get_all_assets_id_by_node_key(org_id=self.org_id, node_key=self.key)
+        return set(_assets_id)
+
+    @classmethod
+    def get_all_assets_id_by_node_key(cls, org_id, node_key):
+        _mapping = cls.get_node_all_assets_id_mapping(org_id)
```
Comment (ibuler) [inline apps/assets/models/node.py:348]: nodekey_assetsid_mapping

File: apps/assets/models/node.py [hunk 4]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
+            cls.set_node_all_assets_id_mapping_to_memory(org_id, mapping=_mapping)
+            return _mapping
+
+        return {}
+
+    # from memory
+
+    @classmethod
+    def get_node_all_assets_id_mapping_from_memory(cls, org_id):
+        _mapping = cls.org_mapping_to_node_all_assets_id_mapping[org_id]
```
Comment (ibuler) [inline apps/assets/models/node.py:371]: cls.org_mapping_to_node_all_assets_id_mapping.get(org_id, {})

File: apps/assets/models/node.py [hunk 5]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
+        cache.set(cache_key, mapping, timeout=None)
+
+    @classmethod
+    def expire_node_all_assets_id_mapping_from_cache(cls, org_id):
+        cache_key = cls._get_cache_key_for_node_all_assets_id_mapping(org_id)
+        cache.delete(cache_key)
+
+    @staticmethod
+    def _get_cache_key_for_node_all_assets_id_mapping(org_id):
+        return 'CACHE_KEY_FOR_ORG_{}_NODE_ALL_ASSETS_ID_MAPPING_OF_ORG'.format(org_id)
```
Comment (ibuler) [inline apps/assets/models/node.py:433]: ASSETS_ORG_NODE_ALL_ASSETS_ID_MAPPING_{}

File: apps/assets/models/node.py [hunk 6]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
+
+    # to generate
+
+    @classmethod
+    def generate_node_all_assets_id_mapping_with_lock(cls, org_id):
```
Comment (ibuler) [inline apps/assets/models/node.py:438]: 使用一个锁

File: apps/assets/models/node.py [hunk 7]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
+        _mapping = {}
+
+        # 如果10s未生成，那么其他生成同一组织下数据的进程进来，也会生成, 锁机制将失去意义
+        lock_key = f'KEY_LOCK_GENERATE_ORG_{org_id}_NODE_ALL_ASSETS_ID_MAPPING'
+        lock = cache.lock(lock_key, expire=10)
+        if not lock.acquire(timeout=1):
+            logger.debug('No acquire lock for `{}` -> pid={}'.format(lock_key, os.getpid()))
```
Comment (ibuler) [inline apps/assets/models/node.py:445]: Not acquire generate org node all assets id lock: key={} pid={}

File: apps/assets/models/node.py [hunk 8]
```diff
@@ -303,10 +325,172 @@ def get_nodes_all_assets(cls, nodes_keys, extra_assets_ids=None):
...
+        for node_id, asset_id in nodes_assets_id:
+            if node_id not in node_id_ancestor_keys_mapping:
+                continue
+            node_ancestor_keys = node_id_ancestor_keys_mapping[node_id]
+            for ancestor_key in node_ancestor_keys:
+                _mapping[ancestor_key].add(asset_id)
+
+        t3 = time.time()
+
+        print('t1-t2(DB Query): {} s, t3-t2(Generate mapping): {} s'.format(t2-t1, t3-t2))
```
Comment (ibuler) [inline apps/assets/models/node.py:491]: debug 用 logger 来

File: apps/assets/models/node.py [hunk 9]
```diff
@@ -1,26 +1,33 @@
...
 from orgs.utils import get_current_org, tmp_to_org
 from orgs.models import Organization
 
 
-__all__ = ['Node', 'FamilyMixin', 'compute_parent_key']
+__all__ = ['Node', 'FamilyMixin', 'compute_parent_key', 'NodeQuerySet']
 logger = get_logger(__name__)
 
 
+def expression_wrapper_to_char_field(field_name):
```
Comment (ibuler) [inline apps/assets/models/node.py:27]: as_string

File: apps/assets/models/node.py [hunk 10]
```diff
@@ -247,9 +256,138 @@ def get_family(self):
...
+
+        # 每秒从cache中获取一次
+        mapping = {}
+        wait_for_seconds = timeout + 1
+        for i in range(wait_for_seconds):
+            mapping = cache.get(cache_key)
+            if mapping:
+                break
+            time.sleep(1)
+        return mapping
```
Comment (BaiJiangJie) [inline apps/assets/models/node.py:331]: 删除timeout相关代码

**Review by ibuler — APPROVED** [review]

**Pull request merged**